karabo_data/geometry2.py

```diff
@@ -631,6 +737,228 @@ def plot_data(self, modules_data):
+        flat, aligned and equally spaced within their quadrant.
+
+        The quadrant positions refer to the corner of each quadrant
+        where module 4, tile 16 is positioned.
+        This is the corner of the last pixel as the data is stored.
+        In the initial detector layout, the corner positions are for the top
+        left corner of the quadrant, looking along the beam.
+
+        The origin of the coordinates is in the centre of the detector.
+        Coordinates increase upwards and to the left (looking along the beam).
+
+        Parameters
+        ----------
+        quad_pos: list of 2-tuples
+          (x, y) coordinates of the last corner (the one by module 4) of each
+          quadrant.
+        unit: float, optional
+          The conversion factor to put the coordinates into metres.
+          The default 1e-3 means the numbers are in millimetres.
+        asic_gap: float, optional
+          The gap between adjacent tiles/ASICs. The default is 4 pixels.
+        panel_gap: float, optional
+          The gap between adjacent modules/panels. The default is 4 pixels.
+        """
+        px_conversion = unit / cls.pixel_size
+        asic_gap_px = 4 if (asic_gap is None) else asic_gap * px_conversion
+        panel_gap_px = 4 if (panel_gap is None) else panel_gap * px_conversion
+
+        tile_size = np.array([cls.frag_fs_pixels, cls.frag_ss_pixels, 0])
+
+        panels_across = [-1, -1, 0, 0]
+        panels_up = [0, -1, -1, 0]
+        modules = []
+        for p in range(16):
+            quad = p // 4
+            quad_corner_x = quad_pos[quad][0] * px_conversion
+            quad_corner_y = quad_pos[quad][1] * px_conversion
+
+            p_in_quad = p % 4
+            # Top beam-left corner of panel
+            panel_corner_x = (quad_corner_x +
+                              (panels_across[p_in_quad] * (
+                                          256 + asic_gap_px + panel_gap_px)))
+            panel_corner_y = (quad_corner_y +
+                              (panels_up[p_in_quad] * (256 + (
+                                          7 * asic_gap_px) + panel_gap_px)))
+
+            tiles = []
+            modules.append(tiles)
+
+            for a in range(16):
+                if a < 8:
+                    up = -a
+                    across = -1
+                else:
+                    up = -(15 - a)
+                    across = 0
+
+                tile_last_corner = (
+                    np.array([panel_corner_x, panel_corner_y, 0.])
+                    + np.array([across, 0, 0]) * (cls.frag_fs_pixels + asic_gap_px)
+                    + np.array([0, up, 0]) * (cls.frag_ss_pixels + asic_gap_px)
+                )
+                tile_first_corner = tile_last_corner - tile_size
+
+                tiles.append(GeometryFragment(
+                    corner_pos=tile_first_corner,
+                    ss_vec=np.array([0, 1, 0]),
```

Comment:
It's definitely the intention to store the geometry with sub-pixel accuracy so we can pass it to tools like CrystFEL or PyFAI (we haven't yet written code to export those formats for LPD, but it will come).

We could also implement `position_modules_interpolate`, but it's *really* slow - the [example from AGIPD](https://karabo-data.readthedocs.io/en/latest/agipd_geometry.html) takes ~7 seconds to assemble one image. And interpolating blurs the image, because each pixel of the result is affected by several input pixels.

If people want to do accurate analysis of detector data, the only practical way is to bring accurate geometry information into the analysis tools (like PyFAI) and give them un-assembled data. Assembling images is only really suitable for a visual inspection of the data.